alacritty_terminal/src/url.rs

```diff
@@ -1,318 +1,40 @@
-// Copyright 2016 Joe Wilm, The Alacritty Project Contributors
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//     http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
+use crate::ansi::TermInfo;
+use crate::index::{Column, Linear, Point};
+use crate::term::Term;
+use std::ops::RangeInclusive;
 
-use unicode_width::UnicodeWidthChar;
-
-use crate::term::cell::{Cell, Flags};
-
-// See https://tools.ietf.org/html/rfc3987#page-13
-const URL_SEPARATOR_CHARS: [char; 10] = ['<', '>', '"', ' ', '{', '}', '|', '\\', '^', '`'];
-const URL_DENY_END_CHARS: [char; 7] = ['.', ',', ';', ':', '?', '!', '('];
-const URL_SCHEMES: [&str; 8] =
-    ["http://", "https://", "mailto:", "news:", "file://", "git://", "ssh://", "ftp://"];
-
-/// URL text and origin of the original click position.
-#[derive(Debug, PartialEq)]
 pub struct Url {
```

Comment:
Would you mind adding `#derive(Debug)` (at least) here?

Reply:
Why? I usually like to leave out derives when they aren't necessary, since they always have a decent impact on compile time. And currently this isn't exported publicly either, which is usually a big argument for adding all the possible derives.

If it's about debugging issues, it shouldn't be a problem to add that while debugging. But it's never printed anywhere, so a debug output isn't really necessary I think.

Reply:
On my system system the build times with `#[derive(Debug)]` and without it are identical ( in debug mode, which I'm usually running to test usability PRs ). I was just adding this macro anyway, while testing the PR.

If you're about release builds, it's about the way rust is doing `lto` and how it calls linker ( and the choice of the linker ) with this target triple `x86_64-pc-linux-gnu`.

It's up to you though.

Reply:
> On my system system the build times with #[derive(Debug)] and without it are identical ( in debug mode, which I'm usually running to test usability PRs ). I was just adding this macro anyway, while testing the PR.

Oh yeah, being cautious with adding a lot of derives will only really make a significant impact when you have a ton or complex derives (or both).

Honestly if there was a `derive(Debug)` statement and you would have asked for its removal, I would have likely responded in a similar fashion. In general I just don't think worrying too much about derives makes a lot of sense unless the type is involved in a public API.

Really the primary reason why I'm still considering this is that with the library split, we'll likely have the `Url` as part of `alacritty_terminal`'s API, in which case it would be a good idea to provide as many derives as possible.